### PR TITLE
Add CSS properties

### PR DESCRIPTION
**Pull request by torusrxxx**
Add CSS properties isolation,object-position,pointer-events,rotate,row-gap

### CODE REVIEW

File: src/freenet/client/filter/CSSTokenizerFilter.java [hunk 1]
```diff
@@ -1212,6 +1214,10 @@ else if ("nav-down".equalsIgnoreCase(element)) {
 			elementVerifiers.put(element, new CSSPropertyVerifier(Arrays.asList("contain","cover","fill","none","scale-down"), ElementInfo.VISUALMEDIA));
 			allelementVerifiers.remove(element);
 		}
+		else if("object-position".equalsIgnoreCase(element)) {
+			elementVerifiers.put(element, new CSSPropertyVerifier(null, ElementInfo.VISUALPAGEDMEDIA,null,Arrays.asList("4a5a40","4a40 5a40","5a40 4a40","4 40 5 40","5 40 4 40")));
```
Comment (ArneBab) [inline src/freenet/client/filter/CSSTokenizerFilter.java:1218]: as far as I can tell the allowed values for `object-position` should be the same as for `background-position`.

- https://developer.mozilla.org/en-US/docs/Web/CSS/object-position#syntax
- https://developer.mozilla.org/en-US/docs/Web/CSS/background-position#syntax

The implementation here is different, though. Can you double-check and harmonize both? Given the complexity of this value, they may need more tests.

File: test/freenet/client/filter/CSSParserTest.java [hunk 2]
```diff
@@ -998,7 +998,9 @@ public class CSSParserTest {
 		propertyTests.put("#x { text-orientation: upright; }", "#x { text-orientation: upright; }");
 		propertyTests.put("#x { scroll-snap-align: start end; }", "#x { scroll-snap-align: start end; }");
 		propertyTests.put("#x { scroll-snap-stop: always; }", "#x { scroll-snap-stop: always; }");
-		propertyTests.put("#x { scroll-snap-type: both proximity; }", "#x { scroll-snap-type: both proximity; }");
+		propertyTests.put("#x { scroll-snap-type: both proximity; } #y { scroll-snap-type: x; }", "#x { scroll-snap-type: both proximity; } #y { scroll-snap-type: x; }");
+		propertyTests.put("#box1:hover {rotate: 90deg;} #box2:hover {rotate: y 180deg;} #box3:hover {rotate: 1 2 1 360deg;} #box4:hover {rotate: none;}", "#box1:hover {rotate: 90deg;} #box2:hover {rotate: y 180deg;} #box3:hover {rotate: 1 2 1 360deg;} #box4:hover {rotate: none;}");
+		propertyTests.put("#a {object-position: top;} #b {object-position: 25% 75%;} #c {object-position: 1cm top;} #d {object-position: bottom 10px right 20px;}", "#a {object-position: top;} #b {object-position: 25% 75%;} #c {object-position: 1cm top;} #d {object-position: bottom 10px right 20px;}");
```
Comment (ArneBab) [inline test/freenet/client/filter/CSSParserTest.java:1003]: It would be nice to have tests here that show that illegal values (ideally edge-cases) are rejected, both for rotate and for object-position.

**Review by ArneBab — CHANGES_REQUESTED** [review]
would be great to harmonize object-position and background-position, and object-position and rotate need some tests that illegal values are rejected.

**Comment by ArneBab** [comment]
Thank you for the PR! These look pretty useful for freesite authors! (and for being able to simply mirror arbitrary pages)

File: src/freenet/client/filter/CSSTokenizerFilter.java (hunk 1 above)
Comment (torusrxxx) [reply src/freenet/client/filter/CSSTokenizerFilter.java:1218]: background-position filter is broken, so I'll just use the new object-position for both.

**Review by ArneBab — APPROVED** [review]
looks good to me — thank you!

**Comment by torusrxxx** [comment]
Should I continue updating more CSS properties here or should I create new branches?

**Comment by torusrxxx** [comment]
40 properties supported in this pull request:

1. `border-block-end`
2. `border-block-start`
3. `border-inline-end`
4. `border-inline-start`
5. `isolation`
6. `object-position`
7. `overflow-block`
8. `overflow-inline`
9. `overflow-x`
10. `overflow-y`
11. `overscroll-behavior`
12. `overscroll-behavior-block`
13. `overscroll-behavior-inline`
14. `overscroll-behavior-x`
15. `overscroll-behavior-y`
16. `pointer-events`
17. `rotate`
18. `row-gap`
19. `scroll-margin`
20. `scroll-margin-block`
21. `scroll-margin-block-end`
22. `scroll-margin-block-start`
23. `scroll-margin-bottom`
24. `scroll-margin-inline`
25. `scroll-margin-inline-end`
26. `scroll-margin-inline-start`
27. `scroll-margin-left`
28. `scroll-margin-right`
29. `scroll-margin-top`
30. `scroll-padding`
31. `scroll-padding-block`
32. `scroll-padding-block-end`
33. `scroll-padding-block-start`
34. `scroll-padding-bottom`
35. `scroll-padding-inline`
36. `scroll-padding-inline-end`
37. `scroll-padding-inline-start`
38. `scroll-padding-left`
39. `scroll-padding-right`
40. `scroll-padding-top`

163 currently still not supported new CSS properties:

1. `accent-color`
2. `alignment-baseline`
3. `all`
4. `animation`
5. `animation-composition`
6. `animation-delay`
7. `animation-direction`
8. `animation-duration`
9. `animation-fill-mode`
10. `animation-iteration-count`
11. `animation-name`
12. `animation-play-state`
13. `animation-timing-function`
14. `appearance`
15. `aspect-ratio`
16. `backdrop-filter`
17. `backface-visibility`
18. `background-position-x`
19. `background-position-y`
20. `border-block`
21. `border-block-color`
22. `border-block-style`
23. `border-block-width`
24. `border-inline`
25. `border-inline-color`
26. `border-inline-style`
27. `border-inline-width`
28. `clip-path`
29. `clip-rule`
30. `color-interpolation-filters`
31. `contain`
32. `contain-intrinsic-block-size`
33. `contain-intrinsic-height`
34. `contain-intrinsic-inline-size`
35. `contain-intrinsic-size`
36. `contain-intrinsic-width`
37. `container`
38. `container-name`
39. `container-type`
40. `content-visibility`
41. `counter-set`
42. `fill`
43. `fill-opacity`
44. `fill-rule`
45. `filter`
46. `flood-color`
47. `flood-opacity`
48. `font-feature-settings`
49. `font-language-override`
50. `font-optical-sizing`
51. `font-palette`
52. `font-size-adjust`
53. `font-smooth Non-standard`
54. `font-synthesis`
55. `font-synthesis-small-caps`
56. `font-synthesis-style`
57. `font-synthesis-weight`
58. `font-variant-alternates`
59. `font-variant-caps`
60. `font-variant-east-asian`
61. `font-variant-emoji`
62. `font-variant-ligatures`
63. `font-variant-numeric`
64. `font-variant-position`
65. `font-variation-settings`
66. `forced-color-adjust`
67. `gap`
68. `grid`
69. `grid-area`
70. `grid-auto-columns`
71. `grid-auto-flow`
72. `grid-auto-rows`
73. `grid-column`
74. `grid-column-end`
75. `grid-column-start`
76. `grid-row`
77. `grid-row-end`
78. `grid-row-start`
79. `grid-template`
80. `grid-template-areas`
81. `grid-template-columns`
82. `grid-template-rows`
83. `hyphenate-character`
84. `hyphenate-limit-chars`
85. `hyphens`
86. `image-orientation`
87. `image-rendering`
88. `initial-letter`
89. `inset`
90. `inset-block`
91. `inset-block-end`
92. `inset-block-start`
93. `inset-inline`
94. `inset-inline-end`
95. `inset-inline-start`
96. `line-clamp`
97. `marker`
98. `marker-end`
99. `marker-mid`
100. `marker-start`
101. `mask`
102. `mask-border`
103. `mask-border-mode`
104. `mask-border-outset`
105. `mask-border-repeat`
106. `mask-border-slice`
107. `mask-border-source`
108. `mask-border-width`
109. `mask-clip`
110. `mask-composite`
111. `mask-image`
112. `mask-mode`
113. `mask-origin`
114. `mask-position`
115. `mask-repeat`
116. `mask-size`
117. `mask-type`
118. `math-depth`
119. `offset`
120. `offset-anchor`
121. `offset-distance`
122. `offset-path`
123. `offset-position`
124. `offset-rotate`
125. `overflow-anchor`
126. `overflow-clip-margin`
127. `page`
128. `paint-order`
129. `perspective`
130. `perspective-origin`
131. `place-content`
132. `place-items`
133. `place-self`
134. `print-color-adjust`
135. `ruby-align`
136. `ruby-position`
137. `scale`
138. `scrollbar-color`
139. `scrollbar-gutter`
140. `scrollbar-width`
141. `shape-image-threshold`
142. `shape-margin`
143. `shape-outside`
144. `stop-color`
145. `stop-opacity`
146. `text-anchor`
147. `text-box`
148. `text-box-edge`
149. `text-box-trim`
150. `text-decoration-skip-ink`
151. `text-rendering`
152. `text-wrap-mode`
153. `text-wrap-style`
154. `touch-action`
155. `transform-box`
156. `transform-style`
157. `transition`
158. `transition-behavior`
159. `translate`
160. `user-select`
161. `vector-effect`
162. `view-transition-name`
163. `will-change`

31 currently supported non-standard or deprecated CSS properties needing review:

1. `azimuth`
2. `box-suppress`
3. `clip`
4. `color-rendering`
5. `cue`
6. `cue-after`
7. `cue-before`
8. `elevation`
9. `nav-down`
10. `nav-left`
11. `nav-right`
12. `nav-up`
13. `pause`
14. `pause-after`
15. `pause-before`
16. `pitch`
17. `pitch-range`
18. `play-during`
19. `punctuation-trim`
20. `richness`
21. `speak`
22. `speak-header`
23. `speak-numeral`
24. `speak-punctuation`
25. `speech-rate`
26. `stress`
27. `text-autospace`
28. `text-decoration-skip`
29. `text-outline`
30. `voice-family`
31. `volume`

**Comment by ArneBab** [comment]
Normally I’d say "please create new branches if a PR is already approved" (I go through the PRs every few weekends and merge what’s approved, so if you add another commit after approval, I cannot merge the PR), but if the commits are obviously correct, I’ll review them by commit.

File: src/freenet/client/filter/CSSTokenizerFilter.java [hunk 3]
```diff
@@ -1101,18 +1123,22 @@ else if("list-style".equalsIgnoreCase(element))
...
-				|| "margin-block-end".equalsIgnoreCase(element) || "margin-block-start".equalsIgnoreCase(element) || "margin-inline-end".equalsIgnoreCase(element) || "margin-inline-start".equalsIgnoreCase(element))
+				|| "margin-block-end".equalsIgnoreCase(element) || "margin-block-start".equalsIgnoreCase(element) || "margin-inline-end".equalsIgnoreCase(element) || "margin-inline-start".equalsIgnoreCase(element)
+				|| "scroll-padding-right".equalsIgnoreCase(element) || "scroll-padding-left".equalsIgnoreCase(element) || "scroll-padding-top".equalsIgnoreCase(element) || "scroll-padding-bottom".equalsIgnoreCase(element)
+				|| "scroll-padding-block-end".equalsIgnoreCase(element) || "scroll-padding-block-start".equalsIgnoreCase(element) || "scroll-padding-inline-end".equalsIgnoreCase(element) || "scroll-padding-inline-start".equalsIgnoreCase(element))
 		{
 			//margin-width=Length|Percentage|Auto
+			// scroll-padding-* can have value auto, so it is the same as margin-*; scroll-margin-* can't have value auto, so it is the same as padding-*
```
Comment (ArneBab) [inline src/freenet/client/filter/CSSTokenizerFilter.java:1172]: thank you for this comment!

File: src/freenet/client/filter/CSSTokenizerFilter.java [hunk 4]
```diff
@@ -82,24 +82,6 @@ public boolean isValidURI(String URI)
...
-			return b;
-		}
-		if (blen == 0) {
-			return a;
-		}
-		@SuppressWarnings("unchecked") final T[] result = (T[]) java.lang.reflect.Array.newInstance(a.getClass().getComponentType(), alen + blen);
-		System.arraycopy(a, 0, result, 0, alen);
-		System.arraycopy(b, 0, result, alen, blen);
-		return result;
-	}
```
Comment (ArneBab) [inline src/freenet/client/filter/CSSTokenizerFilter.java:100]: Careful: this is public, so it could be used by plugins ⇒ unused info doesn’t tell us much. That’s annoying, but fixing it will take a long time.

(we need a clearly specified plugin API to avoid this, but that currently does not exist, even though there’s the HighLevelSimpleClient, because plugins can access the full node)

⇒ please restore the function, otherwise plugins we don’t know may break.

**Review by ArneBab — CHANGES_REQUESTED** [review]
Change needed: https://github.com/hyphanet/fred/pull/1050/commits/1e6c4bd23150a8c63d4445da34a2030c0190064c#r2072343365

File: src/freenet/client/filter/CSSTokenizerFilter.java [hunk 5]
```diff
@@ -510,11 +529,20 @@ private static void addVerifier(String element)
 		} else if("align-self".equalsIgnoreCase(element)) {
 			elementVerifiers.put(element, new CSSPropertyVerifier(Arrays.asList("auto", "stretch", "baseline", "last-baseline"), ElementInfo.VISUALMEDIA, null, Arrays.asList("127"), true, true));
 			allelementVerifiers.remove(element);
+		} else if("all".equalsIgnoreCase(element)) {
+			elementVerifiers.put(element, new CSSPropertyVerifier(null, ElementInfo.MEDIA, null, null, true, true));
```
Comment (ArneBab) [inline src/freenet/client/filter/CSSTokenizerFilter.java:535]: In line 3758 there’s a comment that notes only `"initial", "inherit" and "unset"` — can you fix that to also name "revert" and "revert-layer"?

That way it’s also obvious from a quick look at the code why "all" here matches the definition in https://developer.mozilla.org/de/docs/Web/CSS/all

File: src/freenet/client/filter/CSSTokenizerFilter.java [hunk 6]
```diff
@@ -1450,6 +1527,7 @@ else if("text-decoration-line".equalsIgnoreCase(element))
 		}
 		else if("text-decoration-skip".equalsIgnoreCase(element))
 		{
+			// FIXME: the specified values here are invalid
```
Comment (ArneBab) [inline src/freenet/client/filter/CSSTokenizerFilter.java:1532]: were they valid once (⇒ are they accepted by browsers)? If yes, we shouldn’t delete them, because a mirrored page or an old page may (still) use them. Some of the freesites are 15 years old and can’t be updated anymore but are still useful references.

**Comment by ArneBab** [comment]
I now reviewed all the commits — that’s great work! Thank you!

If you manage to finish updating the filter, I’d guess that there’ll then be few people with a similarly complete knowledge about available CSS options as you’ll then have …

File: src/freenet/client/filter/CSSTokenizerFilter.java (hunk 6 above)
Comment (torusrxxx) [reply src/freenet/client/filter/CSSTokenizerFilter.java:1532]: None of these are accepted by Firefox or Chrome. The value "spaces" and may be "ink" may have worked for a short time in Chrome, but this entire property has been unrecognized by Firefox and Chrome now.

File: src/freenet/client/filter/CSSTokenizerFilter.java (hunk 4 above)
Comment (torusrxxx) [reply src/freenet/client/filter/CSSTokenizerFilter.java]: OK, re-added and marked Deprecated.

File: src/freenet/client/filter/CSSTokenizerFilter.java (hunk 6 above)
Comment (ArneBab) [reply src/freenet/client/filter/CSSTokenizerFilter.java:1552]: :+1:

**Review by ArneBab — APPROVED** [review]
Looks good!

**Comment by torusrxxx** [comment]
Thank you. I guess I am gonna watch videos now, 58 new CSS properties might be too many for a single PR!

**Comment by ArneBab** [comment]
Taking a break is important ☺
